### ISSUE TITLE
Ask some questions about keqing and fischl in gcsim

### ISSUE THREAD

**Issue by NRZ091** [genshinsim/gcsim#1400]
Because English is not proficient, translation is used, which may lead to inaccurate translation of some sentences
         I wonder if keqing's skill in gcsim uses long press by default.(Of course, there may also be a bug in the damage frame of Keqing's skill  in gcsim)
        That's when I found keqing using skll, skill, attack, burst.
        ●gcsim is simulated as
skill(trigger little Amy A4),skill(trigger little Amy A4), attack(do not trigger when the injury interval from E is 0.44s).
        ● Actual combat simulation
skill(trigger little Amy A4),skill(not trigger),attack(trigger)
        Although they all swallow one and have little impact on the total injury, is it because gcsim Sunny eE uses long press e? 
(˘•ω•˘)
        Because according to the theoretical time of gcsim, eE can trigger little Amy A4, the general hand operation time will be longer than the theoretical time, so it is easier to play double trigger, but the actual eE can not play double trigger, so I would like to ask the specific reason and any solution
        gcsim Team Link:
https://gcsim.app/v3/viewer/share/d64d7abc-3a28-4222-82ec-9ad96a9a3338
====
The following is the original Chinese text

想问问gcsim中阿晴的ee是不是默认使用长按e了(当然也有可能是gcsim中阿晴ee的伤害帧有bug)
(ps：我去翻文档的时候发现好像没有区分长e和短e)
gcsim队伍链接：
https://gcsim.app/v3/viewer/share/0cc3481b-a878-4e42-82da-d9a442b6a031

就是我今天在会去对照debug研究刻皇的附着的时候，发现阿晴eEaq时
●gcsim中模拟为
e(触发小艾米A4),E(触发小艾米A4),a(距离E出伤间隔为0.44s不触发)。
●实战模拟为
e(触发小艾米A4),E(不触发),a(触发)

虽然都会吞一个而且对总伤上影响不大，但这里是不是因为gcsim中阿晴eE使用了长按e呢？( ˘•ω•˘ ) 
因为按照gcsim的理论时间来说eE都能触发小艾米A4，一般手操的时间会大于理论时间，从而更容易打出双触发才对，但实际eE打不出来双触发，所以想问问具体原因和有没有什么解决办法
●ｖ●

**Comment by k0l11** [comment]
Yes, this is bugged. Keqing's hitlag on E is supposed to happen on the "Stellar Restoration (Slashing)" dmg instead of "Stellar Restoration" dmg.

**Comment by k0l11** [comment]
Note: There is no `hold` param for Keqing Hold E so `skill[hold=5]` is the same as `skill`.

**Comment by NRZ091** [comment]
Thank you very much

(ps:
"skill[hold=5]" is me trying to see if there is a difference between long and short e, and maybe I forgot to delete it later when I shared the link)

In addition, when I just checked debug, I found that the triggering relationship between keqing and fischl A4 in gcsim web was adjusted, but it was still the same as before

For example
keqing skill(fischl A4 is triggered),skill(fischl A4 is triggered),attack(fischl A4 is not triggered)

![image]( #https://github.com/genshinsim/gcsim/assets/108729837/4daf2e86-9bbc-41b5-bdd6-a3930c1b4eca)
![image](https://github.com/genshinsim/gcsim/assets/108729837/69ff043b-941b-4acb-9509-ece68b11e35d)
![image](https://github.com/genshinsim/gcsim/assets/108729837/7ca1bb33-c2ab-477e-85b0-38cbc504d77c)

I don't know why. Sorry

**Comment by k0l11** [comment]
When the next version of gcsim is released (https://github.com/genshinsim/gcsim/releases) it will be live on web.

**Comment by NRZ091** [comment]
> When the next version of gcsim is released (https://github.com/genshinsim/gcsim/releases) it will be live on web.

fischl A4 can now be triggered three times during "keqing skill, skill, attack"
As follows
keqing skill(fischl A4 is triggered),skill(fischl A4 is triggered),attack(fischl A4 is triggered)
![image](https://github.com/genshinsim/gcsim/assets/108729837/6d26af58-203a-4278-bc95-0931e8e9bf29)
![image](https://github.com/genshinsim/gcsim/assets/108729837/a8767348-2964-4554-b3dc-c15666eb222d)
![image](https://github.com/genshinsim/gcsim/assets/108729837/6dfb2bce-eafc-491d-8721-29537393af2b)

**Comment by k0l11** [comment]
Seems like a case of https://library.keqingmains.com/evidence/characters/electro/fischl#fischl-a4-not-05s to me.

**Comment by NRZ091** [comment]
I guess this bug might exist
But the description in that test(#fischl-a4-not-05s) is wrong.In fact, according to TCL's actual data, the E1 of Diluc is 24 frames instead of 45 frames,so at e1 after Diluc n1, Fischl A4 will not trigger
(Diluc frame link: https://library.keqingmains.com/characters/pyro/diluc?ability=e&frames=skill&talent=e#frames)
Here are some of my test videos：
(Of these, 532 is the damage number of Fischl A4)

Test on official cloud game：
https://github.com/genshinsim/gcsim/assets/108729837/d6400ca6-6d6e-44d0-b491-bafd8be41dc3
Test on mobile phone：
https://github.com/genshinsim/gcsim/assets/108729837/5bc3103a-7ec8-431e-a154-339394169b80
https://github.com/genshinsim/gcsim/assets/108729837/e2ca4c67-ea53-4c0c-8ab9-8d76e8736675
In case of slowing down due to the influence of ice adhesion:
(keqing can still only trigger Fischl A4 twice while being slowed down)
https://github.com/genshinsim/gcsim/assets/108729837/447c0734-2e8e-4538-b0b4-329b73217cb3

In the above video, keqing can only trigger Fischl A4 twice each time during the course of "skill, skill, attack".
And the following is true each time
keqing skill(fischl A4 is triggered),skill(fischl A4 is not triggered),attack(fischl A4 is triggered)

But I couldn't test with mouse macros, so I guess there might be a bug in my tests or in the simulation of gcsim
ｍ(o・ω・o)ｍ

**Comment by NRZ091** [comment]
The link to the video on github didn't seem to last long, so I added another one
[测试文件.zip](https://github.com/genshinsim/gcsim/files/11984278/default.zip)

**Comment by NRZ091** [comment]
●I took a look at the relevant section of keqing in gcsim Docs and the related test videos.
(https://docs.gcsim.app/reference/characters/keqing)
It was then discovered that the video only tested the action frame when keqing's skill was released into the void.
This means that in gcsim simulation, there is still some flying time between keqing's two skills.   (In my test video, keqing was used close to target, so the damage interval was shorter than it was in docs.)
This is probably why both keqing's first and second skill triggers Fischl A4 in gcsim.

●In addition, after the update (details can be seen in the two screenshots above)
The action frame for keqing's second skill was adjusted (511F→506F), but the action frame for keqing's attack was not adjusted (it was still 553F).   So the damage interval between keqing's second skill and attack has been extended to more than 30 frames.
So after the adjustment, Fischl A4 will be triggered three times during "keqing skill,skill,attack".

●The following methods may make the simulation of gcsim more accurate:
Add the case when keqing is close to the target, or make the release distance of keqing's skill adjustable by the user
(I guess it would be a bit of a hassle to implement distance adjustment, (and probably create more bugs), and it probably won't be used very often.)
Also, I didn't have the ability to test keqing's detailed motion frame when she was close to the target

So this bug fix may have to be put on hold for a while

**Comment by k0l11** [comment]
https://github.com/genshinsim/gcsim/pull/1594 this should fix all outstanding issues, thanks for the bug report.

**Comment by NRZ091** [comment]
> yer/enemy away, so that the C1 Start damage does not steal your A4 proc opportunity during the 2nd skill.

Thank you very much!

**Comment by NRZ091** [comment]
The relationship between "Keqing skill,skill,attack" and "Fischl A4" should be fine.
●But keqing's other action frames may have forgotten to adjust.
For example：

●In the screenshot that was first uploade (https://github.com/genshinsim/gcsim/issues/1400#issuecomment-1613507939)：
The action frame of keqing's second skill is 511, and the action frame of keqing's attack is 553. The gap between them is 42 frames.
(The interval between them is also 42 frames in KQM TCL,  link:https://library.keqingmains.com/characters/electro/keqing?frames=skill)

●In the screenshot that was second uploaded (https://github.com/genshinsim/gcsim/issues/1400#issuecomment-1623163625)：
The action frame of keqing's second skill is 506, but the action frame of keqing's attack is still 553.The gap between them is 47 frames.

●now：
![image](https://github.com/genshinsim/gcsim/assets/108729837/1f000966-12d2-44e9-8fac-1d5c30a95390)
The action frame of keqing's second skill is 506，but the action frame of keqing's attack is still 553.

So I guess some of keqing's action frames forgot to adjust together


> [This](https://github.com/genshinsim/gcsim/pull/1594) should fix all outstanding issues. Thanks for the bug report. After the fix you will only have 2 Fischl A4 procs during `keqing skill, skill, attack;` at both C0 and C1 unless you add in manual `wait(x);` between the two skills. In C1's case, you will also need to teleport the player/enemy away, so that the C1 Start damage does not steal your A4 proc opportunity during the 2nd skill.

**Comment by k0l11** [comment]
But that should be accurate. The first screenshot having the second skill at 511 instead of 506 is wrong in general because there the hitlag was wrong.

**Comment by k0l11** [comment]
As far as I can tell, there should be no difference in the actual action frames when it comes to teleport distance from a quick check in-game, so it is accurate that if the Slashing/C1 End DMG procs Fischl A4, then Keqing N1 does as well.

**Comment by NRZ091** [comment]
I Get It.
There are some issues in KQM TCL.
In KQM TCL( https://docs.google.com/spreadsheets/d/1l5DOZ6RgYYMIxMtJtd7oQNL9WWVNXcQL0nwan7q7QGc/edit#gid=0 ), hitlag is labeled "E Thunderclap Slash" instead of "E Recast", so I assumed that E Recast was not affected by hitlag

Thank you very much

**Comment by k0l11** [comment]
Looks like the thing about being 1f too short is a more systemic issue since many other chars have this behavior if their E/Q has hitlag and I can't figure out what's causing it. I fixed the Recast E -> Dash thing (15f -> ~22f after), but couldn't fix the 1f off issue. Please check if there is any remaining issue except 1f off. If there isn't one, then this can be closed once #1597 is merged.

**Comment by NRZ091** [comment]
I think there should be no problems for the time being.
emmmm...
Suddenly realized that if Keqing's E Cast and E Recast are used close to the target, not only E Recast's damage frame will be shortened, but also E Cast's damage frame will be shortened.
(It seems a little inappropriate to bring it up now)
But the impact is likely to be relatively small.
Thank you very much anyway

**Comment by k0l11** [comment]
Keqing's E Cast hitmark frames are already done standing close to the target, see https://youtu.be/kI6N3Mn5BQY?t=34.
